extensions/smb/SmbConnectionControllerService.cpp

```diff
@@ -0,0 +1,119 @@
+/**
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "SmbConnectionControllerService.h"
+#include "core/Resource.h"
+#include "utils/OsUtils.h"
+#include "utils/expected.h"
+
+namespace org::apache::nifi::minifi::extensions::smb {
+
+void SmbConnectionControllerService::initialize() {
+  setSupportedProperties(Properties);
+}
+
+void SmbConnectionControllerService::onEnable()  {
+  std::string hostname;
+  std::string share;
+
+  if (!getProperty(Hostname, hostname))
+    throw Exception(PROCESS_SCHEDULE_EXCEPTION, "Missing hostname");
+
+  if (!getProperty(Share, share))
+    throw Exception(PROCESS_SCHEDULE_EXCEPTION, "Missing share");
+
+  server_path_ = "\\\\" + hostname + "\\" + share;
+
+  auto password = getProperty(Password);
+  auto username = getProperty(Username);
+
+  if (password.has_value() != username.has_value())
+    throw Exception(PROCESS_SCHEDULE_EXCEPTION,  "Either both a username and a password, or neither of them should be provided.");
+
+  if (username.has_value())
+    credentials_.emplace(Credentials{.username = *username, .password = *password});
+  else
+    credentials_.reset();
+
+  ZeroMemory(&net_resource_, sizeof(net_resource_));
+  net_resource_.dwType = RESOURCETYPE_DISK;
+  net_resource_.lpLocalName = nullptr;
+  net_resource_.lpRemoteName = server_path_.data();
+  net_resource_.lpProvider = nullptr;
```

Comment:
I know this is the traditional way of initializing a struct on Windows, but I think the C++20 way is nicer:
```suggestion
  net_resource_ = {
    .dwType = RESOURCETYPE_DISK,
    .lpLocalName = nullptr,
    .lpRemoteName = server_path_.data(),
    .lpProvider = nullptr,
  };
```
(it does zero out the rest of the fields)

Reply:
agreed https://github.com/apache/nifi-minifi-cpp/pull/1634/commits/3b513dcd2120d5e5ab361e8230060721450c6551#diff-8f6654a3a37c2021e4cbe35a2c93791afb67776b83389a56a45bf78e3f32bc3eR53-R58